src/adapter/src/rbac.rs

```diff
@@ -169,14 +170,31 @@ pub fn check_plan(
     // Obtain all roles that the current session is a member of.
     let role_membership = catalog.collect_role_membership(role_id);
 
-    // Validate that the current role has the required membership to alter an object's owner.
-    if let Plan::AlterOwner(AlterOwnerPlan { new_owner, .. }) = plan {
-        if !role_membership.contains(new_owner) {
-            return Err(AdapterError::Unauthorized(
-                UnauthorizedError::AlterOwnerMembership {
-                    role_name: catalog.get_role(new_owner).name().to_string(),
-                },
-            ));
+    // Validate that the current session has the required role membership to execute the provided
+    // plan.
+    let required_membership: BTreeSet<_> = generate_required_role_membership(plan)
+        .into_iter()
+        .collect();
+    let unheld_membership: Vec<_> = required_membership.difference(&role_membership).collect();
+    if !unheld_membership.is_empty() {
+        let role_names = unheld_membership
+            .into_iter()
+            .map(|role_id| catalog.get_role(role_id).name().to_string())
+            .collect();
+        return Err(AdapterError::Unauthorized(
+            UnauthorizedError::RoleMembership { role_names },
+        ));
+    }
+
+    // DROP OWNED cannot revoke privileges granted by other roles.
+    if let Plan::DropOwned(DropOwnedPlan { revokes, .. }) = plan {
+        let invalid_revokes: BTreeSet<_> = revokes
+            .drain_filter_swapping(|(_, privilege)| !role_membership.contains(&privilege.grantor))
```

Comment:
I find it odd that a function called `check_` takes a `&mut Plan` because it needs to change the behavior of the plan. Now this function is required to be called to correctly implement this plan, right? That seems wrong, especially since we still have ways to disable rbac. Can this work be done in planning? Can we add the notices needed to be sent into the Plan itself? Why is this a notice instead of an error?

Reply:
> Now this function is required to be called to correctly implement this plan, right? 

No, RBAC being on or off will affect the behavior of `DROP OWNED` but they are both behaviors are correct. When RBAC is on, only revokes that the user has the appropriate privileges will execute. When RBAC is off, all revokes will execute.

> Can this work be done in planning? Can we add the notices needed to be sent into the Plan itself?

It can be done in planning, but we'd have to duplicate the logic of:
- Checking if RBAC is on.
- Checking if the user is a super user.
- Collecting the users role membership.

> Why is this a notice instead of an error?

That's how PostgreSQL behaves.

```
postgres=# CREATE ROLE r1;
CREATE ROLE
postgres=# CREATE TABLE t ();
CREATE TABLE
postgres=# GRANT SELECT ON t TO r1;
GRANT
postgres=# SELECT relacl FROM pg_class WHERE relname = 't';
                  relacl                   
-------------------------------------------
 {postgres=arwdDxt/postgres,r1=r/postgres}
(1 row)

postgres=# SET ROLE r1;
SET
postgres=> DROP OWNED BY r1;
WARNING:  no privileges could be revoked for "t"
WARNING:  no privileges could be revoked for column "tableoid" of relation "t"
WARNING:  no privileges could be revoked for column "cmax" of relation "t"
WARNING:  no privileges could be revoked for column "xmax" of relation "t"
WARNING:  no privileges could be revoked for column "cmin" of relation "t"
WARNING:  no privileges could be revoked for column "xmin" of relation "t"
WARNING:  no privileges could be revoked for column "ctid" of relation "t"
DROP OWNED
postgres=> SELECT relacl FROM pg_class WHERE relname = 't';
                  relacl                   
-------------------------------------------
 {postgres=arwdDxt/postgres,r1=r/postgres}
(1 row)
```
 I think the reasoning is that, even if a user doesn't have the privileges for all side effects, `DROP OWNED` will still do something useful. Otherwise, user A can grant a privilege to user B and then user B is no longer able to execute `DROP OWNED`.

> I find it odd that a function called check_ takes a &mut Plan because it needs to change the behavior of the plan. 

Can this be fixed by coming up with a new name for the function?

Reply:
Ok, that makes sense. But it still appears to me that we have to call this function to do the correct thing. If it's not called (in the RBAC on case), we will drop too many privileges. (Obviously this is contrived because why would we not call this in the on case?)

Can the filtering and notice sending be moved into sequence_drop_owned?

Reply:
(Thinking about it more, I'm now also opposed to this function sending notices.)

Reply:
I'm happy to move it `sequence_drop_owned`, but I think I don't fully understand the motivation. 

> But it still appears to me that we have to call this function to do the correct thing. If it's not called (in the RBAC on case), we will drop too many privileges.

Correct here seems to mean: "prevent the user from doing something without the proper privileges". However, that's true of all statements. For any statement, if we don't call this function when RBAC is turned on, then it will do the incorrect thing.

Reply:
My motivation is that decisions about what to do should not live in a function named check. Those should be made during planning or sequencing.

Reply:
Would renaming the function to something like `check_and_filter` address that or would it still fit better in `sequence_drop_owned`?

Reply:
I still think it'd fit better somewhere else. I like the idea that rbac.rs is in charge of determining if a thing can happen. Other parts can be in charge of sending notices or changing behavior.

Reply:
I moved it to the sequencer.